### ISSUE TITLE
Support GM17 (SE-SNARKs)

### ISSUE THREAD

**Issue by HarryR** [EYBlockchain/zksnark-verifier-standard#3]
Which is now supported by Zokrates.

See:

 * https://github.com/Zokrates/ZoKrates/blob/master/zokrates_core/src/proof_system/bn128/gm17.rs#L193
 * https://github.com/Zoffee/zoffee-onchain/blob/master/contracts/Verifier_GM17.sol

A little more expensive than Groth16(withoutGT), but still less expensive than PGHR13.

**Comment by MichaelConnorOfficial** [comment]
Hi Harry, thanks for taking the time to read through all this, and so quickly!

The verifier_registry and verifier standard in their current form do allow for GM17, as the verifying key is stored as a flattened array, and the verify() function accepts flattened arrays. It’s just that the example implementation only shows pghr13, to give a flavour if the standard. It’d be up to the verifier contract to interpret those flattened arrays correctly in the body of the verify() function, and indeed I have an example implementation for gm17 (which I’ve almost finished the truffle tests for). Said verifier imports a different gm17 structs library instead.

I’m on holiday this week, but I can push it when I’m back at my laptop next week.

I cant currently think of a neat way of implementing a ‘one size fits all’ verify() function which could cope with more than one proof type being passed into it and interpreted correctly.
I can think to do a few things:
- either have a separate verifier contract for each different proof type (simple); or
- handle ‘switch logic’ within the verify() function to switch between cases (gm17 or pghr13 or...) - but I think this is quite ugly; or
- use ‘upgradable proxy contracts’ to progressively add new verify() functions for each new proving system which gets discovered; importing the older functions. Each new function would need a unique name. I have an implementation, but I’m undecided whether I like it.

**Comment by HarryR** [comment]
> either have a separate verifier contract for each different proof type (simple)

Yup this is the way to go. You have a contract which has a `verify` method, which is provided both the verification key, the proof and the inputs.

The registry contract would store the verification keys and verifier contracts (where a verification key, or `vkId`) is associated with specific contract, and a proof specifies a specific `vkId`, then when asked to verify some proof the registry contract can load the correct verification key and pass the user-supplied arguments to the correct contract.
